<a href="https://colab.research.google.com/github/carloshvmoraes/CEB-IoT-5G/blob/main/2022_01_MongoDB_Blockchain.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Preparação do Sistema

## Instalação do MongoDB

Realiza a instalação do MongoDB na máquina virtual e inicia seu serviço no Linux.

In [2]:
!apt install mongodb
!service mongodb start

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  libpcap0.8 libstemmer0d libyaml-cpp0.5v5 mongo-tools mongodb-clients
  mongodb-server mongodb-server-core
The following NEW packages will be installed:
  libpcap0.8 libstemmer0d libyaml-cpp0.5v5 mongo-tools mongodb mongodb-clients
  mongodb-server mongodb-server-core
0 upgraded, 8 newly installed, 0 to remove and 37 not upgraded.
Need to get 53.1 MB of archives.
After this operation, 215 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libpcap0.8 amd64 1.8.1-6ubuntu1.18.04.2 [118 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic/main amd64 libstemmer0d amd64 0+svn585-1build1 [62.5 kB]
Get:3 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libyaml-cpp0.5v5 amd64 0.5.2-4ubuntu1 [150 kB]
Get:4 http://archive.ubuntu.com/ubuntu bionic/universe amd64 mongo-tools amd64 3.6.3-0ubunt

## Classe para Banco de Dados em Blockchain

Realiza o download do código fonte da classe que cria o objeto de banco de dados com blockchain.

In [3]:
!wget https://raw.githubusercontent.com/carloshvmoraes/CEB-IoT-5G/main/2022-01/BancoBlockchain.py -O BancoBlockchain.py

--2022-01-19 21:44:33--  https://raw.githubusercontent.com/carloshvmoraes/CEB-IoT-5G/main/2022-01/BancoBlockchain.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 14093 (14K) [text/plain]
Saving to: ‘BancoBlockchain.py’

BancoBlockchain.py  100%[===================>]  13.76K  --.-KB/s    in 0.1s    

2022-01-19 21:44:37 (139 KB/s) - ‘BancoBlockchain.py’ saved [14093/14093]



# Simulação

A simulação envolve a criação de um conjunto de medidas para teste e construção do banco de dados em blockchain.

## Gerando dados de medição

Cria aleatoriamente um conjunto de medidas definida pelo número de medidas, padrão 10, através de um gerador aleatório com semente fixada em 42.

In [4]:
import random
import datetime 

random.seed(42)

# lista de medidas em bytes
medidas = []

#instante inicial
instante = datetime.datetime(2022, 1, 20,00,00)

num_medidas = 10
for m in range(num_medidas):
  instante = instante + datetime.timedelta(milliseconds=random.randrange(60000))
  id_medidor = random.randint(1,10)
  valor = random.random()*100.0-10.0
  med = {
      'instante':instante.strftime('%Y-%m-%d %H:%M:%S:%f'),
      'id_medidor':id_medidor,
      'valor': valor,
      }
  medidas.append(med)

medidas

[{'id_medidor': 2,
  'instante': '2022-01-20 00:00:41:905000',
  'valor': -7.498924477733306},
 {'id_medidor': 4,
  'instante': '2022-01-20 00:00:59:929000',
  'valor': 12.321073814882276},
 {'id_medidor': 2,
  'instante': '2022-01-20 00:01:48:194000',
  'valor': 57.669948742291126},
 {'id_medidor': 9,
  'instante': '2022-01-20 00:02:46:663000',
  'valor': -1.3061167370583853},
 {'id_medidor': 1,
  'instante': '2022-01-20 00:03:14:314000',
  'valor': -7.020278056192966},
 {'id_medidor': 4,
  'instante': '2022-01-20 00:03:28:642000',
  'valor': 40.53552881033624},
 {'id_medidor': 9,
  'instante': '2022-01-20 00:03:30:381000',
  'valor': 9.88376506866485},
 {'id_medidor': 9,
  'instante': '2022-01-20 00:04:12:971000',
  'valor': 31.95198209616587},
 {'id_medidor': 10,
  'instante': '2022-01-20 00:04:42:410000',
  'valor': 17.819070823066273},
 {'id_medidor': 1,
  'instante': '2022-01-20 00:05:39:380000',
  'valor': 65.88073671297673}]

## Armazenando Dados Operacionais

Cria um banco de dados MongoDB que recebe os blocos em cadeia do blockchain baseados nas medidas obtidas através do gerador aleatório.

In [7]:
from BancoBlockchain import IoTBlockchainDB
from pprint import pprint

banco = IoTBlockchainDB()
banco.reset()

for med in medidas:
  banco.add_transaction(sender=med['id_medidor'],recipient=med['instante'],amount=med['valor'])
  banco.mine_for_next_block()

pprint(banco.get_last_block())
banco.db.command("dbstats")

Bloco #1 adicionado ao blockchain
Bloco #2 adicionado ao blockchain
Bloco #3 adicionado ao blockchain
Bloco #4 adicionado ao blockchain
Bloco #5 adicionado ao blockchain
Bloco #6 adicionado ao blockchain
Bloco #7 adicionado ao blockchain
Bloco #8 adicionado ao blockchain
Bloco #9 adicionado ao blockchain
Bloco #10 adicionado ao blockchain
Bloco #11 adicionado ao blockchain
{'block_reward': 50,
 'difficulty': 1,
 'difficulty_bits': 0,
 'elapsed_time': 3.528594970703125e-05,
 'hash_power': 0.0,
 'height': 11,
 'merkle_root': 'c745f820bbb46dc9172d7d281584aaaf88b325b4ac4cbcb2a7a04f7865944c0d',
 'nonce': 0,
 'number_of_transaction': 2,
 'previous_block': 10,
 'previous_hash': 'e2e9b0e8ef83faa72abb631174626a0bd83a461670aa17c7ac522398c161028c',
 'timestamp': 'Wed Jan 19 21:45:17 2022',
 'transactions': [{'transaction_id': 'e973ca97164e8d0df91eeb25d356e278916662ade766929101197dd419881ca7',
                   'transaction_info': {'amount': 65.88073671297673,
                                    

{'avgObjSize': 758.8181818181819,
 'collections': 1,
 'dataSize': 8347.0,
 'db': 'blockchain',
 'fsTotalSize': 115658190848.0,
 'fsUsedSize': 45369032704.0,
 'indexSize': 4096.0,
 'indexes': 1,
 'numExtents': 0,
 'objects': 11,
 'ok': 1.0,
 'storageSize': 4096.0,
 'views': 0}

O banco de dados ocupou um espaço de 8347 bytes com 11 objetos armazenados, onde são as 10 medidas e o bloco inicial.